rpp30-boc-heartbleed/whimsical
# Frontend workflow

## Description
Installed Recoil for managing state and React Navigation to enable users to navigate between screens. Similar to the backend file structure, each main component has its own folder (inside `/client/components`). Also added a bunch of dummy code in the Container files. I've had a hard time conceptualizing the relationship between the underlying code of a mobile app and tapping/pushing/swiping things on my phone. I included the dummy code to allow for tactile reinforcement for how the app functions and navigates between screens with Navigation (assuming your testing the app on your phone via qr code). Included a bare bones Friends List that uses Recoil state. Hopefully this gives everybody a decent initial feel for how the thing functions.

The state resides in the `/client/state/atoms` folder. There's more on Recoil below. For now I'm thinking each slice of state will have its own file in the atoms folder (without any component specific subfolders). This way we can get a flat, whole picture of our app's state just by opening said folder. Totally understand that this could get messy so happy to adapt and change as needed. Can also create multiple slices of state within a single file, so we've got options re: organization. Would love any feedback on what works and what doesn't as we progress.

Haven't fully dove into the React Native Paper docs yet. I've just been using the StyleSheet component to style within each main component. Every example I've found for React Native styles this way (as well as inline styling) but I'm guessing there'd be a lot less of it by using the prebuilt Paper components.

Haven't hooked up the frontend to the backend yet. I tried to leave as many examples as possible to reduce the learning curve. I'm sure I'm forgetting to mention some things, anything I think of I'll post to slack.

### Frontend Workflow
1. Components can be built out from `/client/components/MyComponent/MyComponentContainer.jsx`. Create new files in the same MyComponent folder for any smaller components.
2. Any additional components (e.g. the Add Errand component) that would be navigated to can be built from the same folder or a new folder, dev's preference. They would also be imported into/exported out of `components/index.js` and imported into `navigation/navigation.jsx`. Add route by creating a new `<Stack.Screen>` component with name and component properties.
3. State values can be added as atoms in `client/state/atoms`. More on Recoil below.
4. `client/common` can house any dummy data for now. I used a js file as an example but we could also create json files with dummy response objects if we wanted to better model API responses.

#### Recoil Primer
Although Recoil has a bunch of cool features, we could probably accomplish most of the MVP by primarily using Atoms. Atoms are single slices of state, objects with a key (a string that must be unique) and a default value. ```
const friendsListState = atom({
  key: 'FriendsListState',
  default: dummyFriendsList,
});
```Using and setting state is refreshingly simple (compared to Redux for instance). Import the Atom and create a hook: `const [myState, setMyState] = useRecoilState(importedState)`. That's it, really. Again there's a lot more that Recoil can do but for maintaining state at the most basic level this is it.

The other main aspect of Recoil is the Selector. Selector's are functions that return a derived state value. They can also act as async functions like fetch requests. We can use a Selector in a component similar to how we use Atoms: `const mySelector = useRecoilValue(myImportedSelector)`.

I also wanted to establish a naming convention to ensure readability. Atom names should end with 'State', basic Selector functions that return derived state should end with 'Selector', and async Selector functions that make an API request should end with 'Query'. I believe maintaining this standard will make the learning experience easier. All that said I still think everyone should refer to the Recoil docs if you have time. Recoil seems to be pretty powerful and I'm only skimming the basics here.

#### Navigation Primer
I've set out examples for how to use Navigation in each components Container file. Hopefully that's enough to get us started but I'd also recommend everybody looking at the Navigation docs as well. It does look to be pretty intuitive though

#### Resources
- [Recoil docs](https://recoiljs.org/docs/introduction/getting-started)
- [React Navigation docs](https://reactnavigation.org/docs/navigating)

## How to test
cd into the client folder and run `expo start`. Scan the qr code with your phone and see that the app renders. Click/swipe around.

## Notes
- `expo install` or `npm install` from the client folder.
- I'll try to connect the frontend and backend by using async Selectors on the client side as soon as possible.
- There doesn't seem to be a lot of options for well maintained Recoil dev tools. I like the look and feel of [Atomos](https://github.com/oslabs-beta/atomos) which has a Chrome extension but I've barely gotten to use it so definitely open to other suggestions.

## Issue tracking
https://trello.com/c/Gd6h9PNd/8-skeleton-code-setup
